pkg/virt-operator/install-strategy/create.go

```diff
@@ -0,0 +1,996 @@
+	for _, sccPriv := range targetStrategy.customSCCPrivileges {
+		var curSccPriv *customSCCPrivilegedAccounts
+		if prevStrategy != nil {
+			for _, entry := range prevStrategy.customSCCPrivileges {
+				if sccPriv.TargetSCC == entry.TargetSCC {
+					curSccPriv = entry
+					break
+				}
+			}
+		}
+
+		privSCCObj, exists, err := stores.SCCCache.GetByKey(sccPriv.TargetSCC)
+		if !exists {
+			continue
+		} else if err != nil {
+			return false, err
+		}
+
+		privSCC, ok := privSCCObj.(*secv1.SecurityContextConstraints)
+		if !ok {
+			return false, fmt.Errorf("couldn't cast object to SecurityContextConstraints: %+v", privSCCObj)
+		}
+		privSCCCopy := privSCC.DeepCopy()
+
+		modified := false
+		users := privSCCCopy.Users
+
+		// remove users from previous
+		if curSccPriv != nil {
+			for _, acc := range curSccPriv.ServiceAccounts {
+				shouldRemove := true
+				// only remove if the target doesn't contain the same
+				// rule, otherwise leave as is.
+				for _, targetAcc := range sccPriv.ServiceAccounts {
+					if acc == targetAcc {
+						shouldRemove = false
+						break
+					}
+				}
+				if shouldRemove {
+					removed := false
+					users, removed = remove(users, acc)
+					modified = modified || removed
+				}
+			}
+		}
+
+		// add any users from target that don't already exist
+		for _, acc := range sccPriv.ServiceAccounts {
+			if !contains(users, acc) {
+				users = append(users, acc)
+				modified = true
+			}
+		}
+
+		if modified {
+			userBytes, err := json.Marshal(users)
+			if err != nil {
+				return false, err
+			}
+
+			data := []byte(fmt.Sprintf(`{"users": %s}`, userBytes))
+			_, err = scc.SecurityContextConstraints().Patch(sccPriv.TargetSCC, types.StrategicMergePatchType, data)
+			if err != nil {
+				return false, fmt.Errorf("unable to patch scc: %v", err)
+			}
+		}
+	}
+
+	if takeUpdatePath {
+		// UPDATE PATH IS
+		// 1. daemonsets - ensures all compute nodes are updated to handle new features
+		// 2. wait for daemonsets to roll over
+		// 3. controllers - ensures controll plane is ready for new features
+		// 4. wait for controllers to roll over
+		// 5. apiserver - toggles on new features.
+
+		// create/update Daemonsets
+		for _, daemonSet := range targetStrategy.daemonSets {
+			err := syncDaemonSet(kv, daemonSet, stores, clientset, expectations)
+			if err != nil {
+				return false, err
+			}
+		}
+
+		// create/update Controller Deployments
+		for _, deployment := range controllerDeployments(targetStrategy) {
+			err := syncDeployment(kv, deployment, stores, clientset, expectations)
+			if err != nil {
+				return false, err
+			}
+
+		}
+
+		// wait for daemonsets
+		for _, daemonSet := range targetStrategy.daemonSets {
+			if !util.DaemonsetIsReady(kv, daemonSet, stores) {
+				log.Log.V(2).Infof("Waiting on daemonset %v to roll over to latest version", daemonSet.GetName())
+				// not rolled out yet
+				return false, nil
+			}
+		}
+		// wait for controller deployments
```

Comment:
Is there something structurally preventing you from putting these waits directly in the controllers?

Reply:
For example, in this scenario you are saying the order is daemonsets, controllerDeployments, then apiDeployments. The controllerDeployments could be configured for the daemonsets to be marked ready and the apiDeployments would be forced to wait for the controllerDeployments to be marked ready.

Reply:
I guess this ordering is really only needed on these upgrades and rollbacks to limit the amount of time unsupported APIs are available.

Reply:
> Is there something structurally preventing you from putting these waits directly in the controllers?

the ordering is only important at the orchestration level depending on the situation. The individual controllers shouldn't need to be aware of this logic.